### ISSUE TITLE
Extern Object Typing Bug

### ISSUE THREAD

**Issue by stp59** [cornell-netlab/petr4#241]
extern object typing is broken when the methods take the object type as args. simple fixes have broken the current typing algorithm so far

**Comment by jnfoster** [comment]
Can you write an example here?

**Comment by hackedy** [comment]
Ok let me procrastinate a bit with an explanation of what's going on here. This came up in the uP4 branch because of the `im_t` extern's `copy_from` method.
```
extern im_t {
  /* ... other methods removed ... */
  void copy_from(im_t im);
}
```
Right now this doesn't typecheck because the typechecker does not insert a forward declaration of `im_t` into the environment before checking the method signatures. Adding code to do this made the declaration typecheck, but causes problems in our type equality/canonicalization code later.

There's a routine `saturate_type` (or something like that) which recursively substitutes type definitions into types. It loops forever when it encounters an `im_t` because the current representation of extern types includes all of its methods. Saturating an `im_t` requires saturating the `copy_from` method, and the `copy_from` method includes a reference to `im_t`, so that has to be saturated as well, etc...

The obvious solution here, I think, is for externs to become strictly nominal[1] and for methods to move into their own context. This could tie in nicely with a refactor to parametrize the typechecker over targets.

[1]: So giving extern types the representation `ExternObjType name`, rather than `ExternObjType {name; methods = [...]}`, with methods kept in a new field of `Prog.CheckerEnv`. The alternative is to introduce true recursive types, which would be a headache considering there's nowhere else in P4 that needs self-reference afaict.

**Comment by hackedy** [comment]
p4c (at git hash c68f3057a) doesn't like it either. I guess the uP4 p4c must be patched to handle this kind of thing properly.
```ryan@larsen-c:~/dev/petr4-aux% p4c example.p4
In file: /home/ryan/dev/p4c/frontends/p4/typeChecking/typeChecker.cpp:168
Compiler Bug: example.p4(3): Could not find type of <Type_Extern>(84) im_t extern im_t<> {
  void set_out_port<>(in PortId_t out_port);
  PortId_t get_in_port<>();
  PortId_t get_out_port<>();
  bit<32> get_value<>(metadata_fields_t field_type);
  void copy_from<>(im_t im);
  void drop<>(); }
extern im_t {
       ^^^^
```

**Comment by hackedy** [comment]
p4lang/p4c#2594

**Comment by hackedy** [comment]
Fixed in #250.
